src/System.Private.CoreLib/src/System/Diagnostics/Eventing/ManagedRuntimeEventSource.cs

```diff
@@ -0,0 +1,115 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+
+using System.Runtime.InteropServices;
+using Microsoft.Win32;
+using Microsoft.Win32.SafeHandles;
+
+using System.Runtime.InteropServices.ComTypes;
+using System.StubHelpers;
+
+
+namespace System.Diagnostics.Tracing
+{
+#if FEATURE_PERFTRACING
+    /// <summary>
+    /// ManagedRuntimeEventSource is an EventSource that represents events emitted by the managed runtime.
+    /// </summary>
+    [EventSource(Name = "System.Runtime")]
+    internal sealed class ManagedRuntimeEventSource : EventSource
+    {
+        // Process-level EventCounters
+        private EventCounter _totalProcessTimeCounter;
+        private EventCounter _workingSetCounter;
+        private EventCounter _virtualMemorySizeCounter;
+        private EventCounter _handleCountCounter;
+        private EventCounter _threadCountCounter;
+
+        // GC Counters
+        private EventCounter _gcTotalMemoryCounter;
+        private EventCounter _gcGen0CollectionCounter;
+        private EventCounter _gcGen1CollectionCounter;
+        private EventCounter _gcGen2CollectionCounter;
+
+        // Exception
+        private EventCounter _exceptionCounter;
+
+        private Timer m_timer;
+
+        private const int EnabledPollingIntervalMilliseconds = 1000; // 1 second
+
+        // Defines the singleton instance for the Resources ETW provider
+        internal static ManagedRuntimeEventSource Log = new ManagedRuntimeEventSource();
+
+        // Threads
+        // TODO
+        public ManagedRuntimeEventSource(): base(EventSourceSettings.EtwSelfDescribingEventFormat)
+        {
+            // TODO: These are all returning fake stuff now
+            _totalProcessTimeCounter = new EventCounter("Total Process Time", this);
+            _workingSetCounter = new EventCounter("Working Set", this);
+            _virtualMemorySizeCounter = new EventCounter("Virtual Memory Size", this);
+            _handleCountCounter = new EventCounter("Handle Count", this);
+            _threadCountCounter = new EventCounter("Thread Count", this);
+
+            // GC counters
```

Comment:
Do you really need to initialize all this during startup here? Can this be initialized lazily, only once somebody asks for the counters?

Reply:
Yeah I should make them to be be initialized lazily. Thanks!